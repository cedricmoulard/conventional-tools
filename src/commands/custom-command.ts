import { Command } from 'commander'
import { CommandData } from '../models/command-data'
import { Option } from '../models/option'
import * as logger from 'npmlog'

export abstract class CustomCommand<T extends CommandData> extends Command {
  protected constructor(name: string, description: string, options: Option[]) {
    super(name)
    this.description(description)

    this.option('-d, --debug <level>', 'output extra debugging (silly, verbose', 'silent')
    this.requiredOption('-p, --preset <preset>', 'preset to use', 'conventionalcommits')
    this.requiredOption('-t, --tag-prefix <prefix>', 'tag prefix to use', '')
    this.requiredOption('-s, --scopes <scopes>', 'conventional commits scope', '*')

    this.action((options: T) => {
      return this.preRun(options)
    })

    options.forEach(option => {
      if (option.required) {
        this.requiredOption(option.flags, option.description, option.defaultValue)
      } else {
        this.option(option.flags, option.description, option.defaultValue)
      }
    })
  }

  private async preRun(options: T): Promise<void> {
    // @ts-ignore
    logger.level = options.debug
    logger.info('[custom-command][preRun]', 'options: %j', options)
    return this.run(options)
  }

  protected abstract run(data: T): Promise<void>
}
