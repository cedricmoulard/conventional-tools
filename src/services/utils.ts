import * as logger from 'npmlog'
import { CommandData } from '../models/command-data'
import { Context, GeneratedContext, Options } from 'conventional-changelog-writer'
import { Commit } from 'conventional-commits-parser'

export const buildPattern = (data: CommandData): RegExp => {
  logger.verbose('[utils][buildPattern]', 'Build pattern with scope "%s"', data.scopes)
  let scopes = data.scopes
  if (!scopes || scopes === '*') {
    scopes = '[\\w\\$\\.\\-\\* ]*'
  }

  const pattern = `^(\\w*)(?:\\((${scopes})\\))?!?\\: (.*)$`
  return new RegExp(pattern)
}
