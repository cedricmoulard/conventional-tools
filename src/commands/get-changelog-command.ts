import * as fs from 'fs'
import * as Stream from 'stream'
import * as logger from 'npmlog'
import { CommandData } from '../models/command-data'
import { CustomCommand } from './custom-command'
import { Option } from '../models/option'
import { getChangelog } from '../services/changelog-service'

export interface GetChangelogData extends CommandData {
  readonly host: string
  readonly config?: string
  readonly outputFile?: string
}

const options: Option[] = [
  {
    flags: '--host <host>',
    description: 'host for link commit',
    defaultValue: '',
    required: true,
  },
  {
    flags: '-n, --config <config-file>',
    description: 'config file in json format. See https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-conventionalcommits',
    required: false,
  },
  {
    flags: '-o, --output-file <file>',
    description: 'write the CHANGELOG to this file',
    required: false,
  },
]

export class GetChangelogCommand extends CustomCommand<GetChangelogData> {
  constructor() {
    super('get-changelog', 'Generate changelog', options)
  }

  protected async run(data: GetChangelogData): Promise<void> {
    const stream = await getChangelog(data)
    let outStream: Stream.Writable = process.stdout

    if (data.outputFile) {
      logger.silly('changelog-command', 'In')
      outStream = fs.createWriteStream(data.outputFile)
    }
    stream.pipe(outStream)
  }
}
