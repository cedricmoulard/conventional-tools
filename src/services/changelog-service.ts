import * as conventionalChangelog from 'conventional-changelog'
import { GetChangelogData } from '../commands/get-changelog-command'
import { buildVersions, getCurrentVersion, getReleaseType } from './versions-services'
import * as logger from 'npmlog'
import { buildPattern } from './utils'
import { Context, GeneratedContext, Options } from 'conventional-changelog-writer'
import { Commit } from 'conventional-commits-parser'

export const RELEASE_COUNT_FOR_EXISTING_TAG = 2
export const RELEASE_COUNT_FOR_CREATE_TAG = 1

const finalize = (host: string, version: string): Options.FinalizeContext<Context, Commit> => {
  return (context: GeneratedContext<Commit, Context>): GeneratedContext => {
    logger.verbose('[utils][finalize]', 'Finalize context')
    logger.silly('[utils][finalize] - input', 'context: %o', context)

    let noteGroups = context.noteGroups.filter(noteGroup => {
      // @ts-ignore
      return noteGroup['notes'].some(note => !!note.commit.type)
    })

    noteGroups = noteGroups.map(noteGroup => {
      // @ts-ignore
      const notes = noteGroup['notes'].filter(note => {
        return !!note.commit.type
      })

      return {
        ...noteGroup,
        notes,
      }
    })

    const newContext = {
      ...context,
      version,
      host,
      commitGroups: context.commitGroups.filter(commitGroup => !!commitGroup.title),
      noteGroups,
    } as GeneratedContext

    logger.silly('[utils][finalize] - output', 'context: %o', newContext)

    return newContext
  }
}

export const getChangelog = async (data: GetChangelogData): Promise<NodeJS.ReadableStream> => {
  logger.verbose('[changelog-service][getChangelog]', 'Enter function')

  const releaseType = await getReleaseType(data)
  const currentVersion = await getCurrentVersion(data)
  const versions = buildVersions(currentVersion, releaseType)

  const { tagExists, tagPrefix, host, preset } = data
  const options: conventionalChangelog.Options = {
    tagPrefix,
    preset,
    releaseCount: tagExists ? RELEASE_COUNT_FOR_EXISTING_TAG : RELEASE_COUNT_FOR_CREATE_TAG,
  }

  logger.verbose('[changelog-service][getChangelog]', 'conventional changelog log options: %o', options)

  return conventionalChangelog(
    options,
    undefined,
    undefined,
    {
      headerPattern: buildPattern(data),
    },
    {
      finalizeContext: finalize(host, data.tagExists ? versions.currentVersion : versions.nextVersion),
    },
  )
}
