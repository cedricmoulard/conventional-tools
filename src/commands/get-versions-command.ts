import { CustomCommand } from './custom-command'
import { CommandData } from '../models/command-data'
import { Option } from '../models/option'
import { getVersions } from '../services/versions-services'
import * as logger from 'npmlog'
import * as fs from 'fs'
import * as Stream from 'stream'

export interface GetVersionData extends CommandData {
  readonly tagPrefix: string
  readonly outputFile?: string
}

const options: Option[] = [
  {
    flags: '-o, --output-file <file>',
    description: 'write the versions to this file',
    required: false,
  },
]

export class GetVersionsCommand extends CustomCommand<GetVersionData> {
  constructor() {
    super('get-versions', 'Get all versions according to conventional commits convention', options)
  }

  protected async run(data: GetVersionData): Promise<void> {
    const stream = await getVersions(data)
    let outStream: Stream.Writable = process.stdout
    if (data.outputFile) {
      logger.silly('changelog-command', 'In')
      outStream = fs.createWriteStream(data.outputFile)
    }
    stream.pipe(outStream)
  }
}
