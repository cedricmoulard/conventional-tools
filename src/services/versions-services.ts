import { Versions } from '../models/versions'
import * as conventionalRecommendedBump from 'conventional-recommended-bump'
import * as gitSemverTags from 'git-semver-tags'
import { clean, inc, ReleaseType } from 'semver'
import { GetVersionData } from '../commands/get-versions-command'
import * as logger from 'npmlog'
import { buildPattern } from './utils'
import BumpOptions = conventionalRecommendedBump.Options
import BumpRecommendation = conventionalRecommendedBump.Callback.Recommendation
import * as str from 'string-to-stream'

const DEFAULT_CURRENT_VERSION = '0.0.0'
const DEFAULT_NEXT_VERSION = '0.1.0'
const DEFAULT_NEXT_MINOR = '0.2.0'
const DEFAULT_NEXT_PATCH = '0.0.1'
const DEFAULT_COMMIT_NUMBER = 0
const MINOR = 'minor'
const PATCH = 'patch'
const RECENT_TAG_INDEX = 0

interface ReleaseInformation {
  releaseType: ReleaseType
  commitNumber: number
}

export const buildVersions = (currentVersion: string, releaseInformation: ReleaseInformation): Versions => {
  logger.verbose('[versions-service][buildVersions]', 'Enter function')
  logger.silly('[versions-service][buildVersions] - input', 'currentVersion: %s', currentVersion)
  logger.silly('[versions-service][buildVersions] - input', 'releaseType: %s', releaseInformation.releaseType)
  logger.silly('[versions-service][buildVersions] - input', 'commitNumber: %s', releaseInformation.commitNumber)

  if (!currentVersion) {
    currentVersion = DEFAULT_CURRENT_VERSION
  }
  const commitNumber = releaseInformation.commitNumber
  const nextRelease = commitNumber > 0 ? inc(currentVersion, releaseInformation.releaseType) || DEFAULT_NEXT_VERSION : currentVersion
  const nextMinor = inc(currentVersion, MINOR) || DEFAULT_NEXT_MINOR
  const nextPatch = inc(currentVersion, PATCH) || DEFAULT_NEXT_PATCH
  const nextCommitTag = nextRelease

  return {
    currentVersion,
    nextRelease,
    nextMinor,
    nextPatch,
    nextCommitTag,
    commitNumber,
  }
}

export function cleanVersion(version: string, tagPrefix = ''): string {
  logger.verbose('[versions-service][cleanVersion]', 'Enter function')
  logger.silly('[versions-service][cleanVersion] - input', 'version: %s', version)
  logger.silly('[versions-service][cleanVersion] - input', 'tagPrefix: %s', tagPrefix)
  const cleanVersion = clean(version.replace(tagPrefix, ''), { loose: true })

  if (!cleanVersion) {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    throw new Error(`Version ${cleanVersion} is not valid`)
  } else {
    return cleanVersion
  }
}

export const getReleaseInformation = (data: GetVersionData): Promise<ReleaseInformation> => {
  logger.verbose('[versions-service][getReleaseType]', 'Enter function')
  logger.silly('[versions-service][getReleaseType] - input', 'data: %j', data)

  let commitNumber = DEFAULT_COMMIT_NUMBER

  const options: BumpOptions = {
    tagPrefix: data.tagPrefix,
    preset: data.preset,
    whatBump: commits => {
      const filteredCommits = commits.filter(commit => !!commit.type)

      const PATCH_LEVEL = 0
      const MINOR_LEVEL = 1
      const MAJOR_LEVEL = 2
      let level = MAJOR_LEVEL

      filteredCommits.forEach(commit => {
        const MERGE_TYPE = 'merge'
        const FEATURE_TYPE = 'feat'
        const { type, notes } = commit

        logger.silly('[versions-service][getReleaseType] - commit', 'type: "%s", notes: "%j"', type, notes)
        let commitLevel = MAJOR_LEVEL

        if (notes && notes.length > 0) {
          commitLevel = PATCH_LEVEL
        } else if (FEATURE_TYPE === commit.type) {
          commitLevel = MINOR_LEVEL
        }

        if (MERGE_TYPE !== commit.type) {
          ++commitNumber
        }

        if (commitLevel < level) {
          level = commitLevel
        }
      })

      return {
        level,
      }
    },
  }

  return new Promise<ReleaseInformation>((resolve, reject) => {
    conventionalRecommendedBump(
      options,
      {
        headerPattern: buildPattern(data),
      },
      (error, recommendation: BumpRecommendation) => {
        if (error) {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
          logger.error('[versions-service][getReleaseType]', error)
          reject(error)
        } else {
          logger.silly('[versions-service][getReleaseType] - output', 'recommendation: %j', recommendation)
          resolve({ releaseType: recommendation.releaseType || 'patch', commitNumber })
        }
      },
    )
  })
}

export const getCurrentVersion = (configuration: gitSemverTags.Options): Promise<string> => {
  return new Promise<string>((resolve, reject) => {
    gitSemverTags(configuration, function (error, tags) {
      if (error) {
        reject(error)
      } else {
        let currentVersion = DEFAULT_CURRENT_VERSION
        if (!!tags && tags.length > 0) {
          currentVersion = cleanVersion(tags[RECENT_TAG_INDEX], configuration.tagPrefix)
        }
        resolve(currentVersion)
      }
    })
  })
}

export const getVersions = async (data: GetVersionData): Promise<NodeJS.ReadableStream> => {
  logger.verbose('[versions-service][getVersions]', 'Enter function')

  const releaseInformation = await getReleaseInformation(data)
  const currentVersion = await getCurrentVersion(data)
  const versions = buildVersions(currentVersion, releaseInformation)

  return str(
    JSON.stringify({
      ...versions,
      nextCommitTag: data.tagPrefix + versions.nextCommitTag,
    }),
  )
}
