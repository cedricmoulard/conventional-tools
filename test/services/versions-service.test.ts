import * as path from 'path'
import * as fs from 'fs-extra'
import { getResultFromStream, prepareRepository } from '../utils'
import { getVersions } from '../../src/services/versions-services'
import { GetVersionData } from '../../src/commands/get-versions-command'

describe('Versions Service', () => {
  const tempPath = path.resolve(__dirname, '..', 'temp', 'versions')

  const globalData: GetVersionData = {
    preset: 'conventionalcommits',
    tagPrefix: '',
    debug: 'silent',
    scopes: '*',
  }

  afterAll(() => {
    fs.removeSync(tempPath)
  })

  describe('given step 1', () => {
    beforeEach(async () => {
      await prepareRepository(tempPath, 'step-1.zip')
    })

    test('returns minor version for module 2', async () => {
      // Given
      const data: GetVersionData = {
        ...globalData,
        tagPrefix: 'module2@',
        scopes: 'module2|shared',
      }

      // When
      const stream = await getVersions(data)
      const actual = await getResultFromStream(stream)
      // Then
      expect(actual).toMatchSnapshot()
    })
  })

  describe('given step 2', () => {
    beforeEach(async () => {
      await prepareRepository(tempPath, 'step-2.zip')
    })

    test('returns patch version for module 2', async () => {
      // Given
      const data: GetVersionData = {
        ...globalData,
        tagPrefix: 'module2@',
        scopes: 'module2|shared',
      }

      // When
      const stream = await getVersions(data)
      const actual = await getResultFromStream(stream)
      // Then
      expect(actual).toMatchSnapshot()
    })
  })

  describe('given step 3', () => {
    beforeEach(async () => {
      await prepareRepository(tempPath, 'step-3.zip')
    })

    test('returns no new version for module 2', async () => {
      // Given
      const data: GetVersionData = {
        ...globalData,
        tagPrefix: 'module2@',
        scopes: 'module2',
      }

      // When
      const stream = await getVersions(data)
      const actual = await getResultFromStream(stream)
      // Then
      expect(actual).toMatchSnapshot()
    })

    test('returns major version for module 2', async () => {
      // Given
      const data: GetVersionData = {
        ...globalData,
        tagPrefix: 'module2@',
        scopes: 'module2|shared',
      }

      // When
      const stream = await getVersions(data)
      const actual = await getResultFromStream(stream)
      // Then
      expect(actual).toMatchSnapshot()
    })

    test('returns no new version for module 1', async () => {
      // Given
      const data: GetVersionData = {
        ...globalData,
        tagPrefix: 'module1@',
        scopes: 'module1|shared',
      }

      // When
      const stream = await getVersions(data)
      const actual = await getResultFromStream(stream)
      // Then
      expect(actual).toMatchSnapshot()
    })
  })
})
