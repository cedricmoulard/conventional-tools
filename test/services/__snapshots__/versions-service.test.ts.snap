// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Versions Service given step 1 returns minor version for module 2 1`] = `"{\\"currentVersion\\":\\"0.0.0\\",\\"nextRelease\\":\\"0.1.0\\",\\"nextMinor\\":\\"0.2.0\\",\\"nextPatch\\":\\"0.0.1\\",\\"commitNumber\\":4}"`;

exports[`Versions Service given step 2 returns patch version for module 2 1`] = `"{\\"currentVersion\\":\\"1.0.0\\",\\"nextRelease\\":\\"1.0.1\\",\\"nextMinor\\":\\"1.1.0\\",\\"nextPatch\\":\\"1.0.1\\",\\"commitNumber\\":1}"`;

exports[`Versions Service given step 3 returns major version for module 2 1`] = `"{\\"currentVersion\\":\\"1.0.1\\",\\"nextRelease\\":\\"2.0.0\\",\\"nextMinor\\":\\"2.1.0\\",\\"nextPatch\\":\\"1.0.2\\",\\"commitNumber\\":1}"`;

exports[`Versions Service given step 3 returns no new version for module 1 1`] = `"{\\"currentVersion\\":\\"2.0.0\\",\\"nextRelease\\":\\"2.0.0\\",\\"nextMinor\\":\\"2.1.0\\",\\"nextPatch\\":\\"2.0.1\\",\\"commitNumber\\":0}"`;

exports[`Versions Service given step 3 returns no new version for module 2 1`] = `"{\\"currentVersion\\":\\"1.0.1\\",\\"nextRelease\\":\\"1.0.1\\",\\"nextMinor\\":\\"1.1.0\\",\\"nextPatch\\":\\"1.0.2\\",\\"commitNumber\\":0}"`;
